blog/_src/posts/2019-05-14-scoping-in-r.md

```diff
@@ -0,0 +1,329 @@
+environment_, i.e. where it is called. In other words, the variable is bound to
+the value last assigned to it (at run time). In our example, just before `f` is
+called, `x` is assigned the value `2`.
+
+Therefore, under dynamic scoping, the example program returns `2`.
+
+Dynamically scoped programming languages include bash and LaTeX. Emacs Lisp is
+dynamically scoped, but allows the programmer to opt into lexical scoping.
+Conversely, Perl and Common Lisp are lexically scoped by default but allow the
+programmer to choose dynamic scoping if desired.
+
+
+## What about R?
+
+The example is written in valid R. If you run it, you will see that the result
+is `1`. Therefore, according to our litmus test, **R is lexically scoped**. But
+is this the whole story?
+
+In _[Evaluating the Design of the R Language][mhov2012]_ (ECOOP 2012)
+([doi][mhov2012-doi]), Morandat, Hill, Osvald, and Vitek write:
+
+> As is often the case, R is lexically scoped up to the point it is not. R
+> is above all a dynamic language with full reflective access to the running
+> program’s data and representation.
+
+[mhov2012]: http://janvitek.org/pubs/ecoop12.pdf
+[mhov2012-doi]: https://doi.org/10.1007/978-3-642-31057-7_6
+
+
+### Circumventing scoping in R
+
+There are many ways to "defeat" scoping in R---made easier because it supports
+explicit manipulation of environments.
+
+Here are some different tricks we can try.
+
+#### Evaluating arbitrary code
+
+```r
+x <- 1
+f <- function(a) {
+  eval(parse(text = "x <- 0"))
+  x + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # 0
+```
+
+On line 3, we parse the string `"x <- 0"` into an unevaluated R expression and
+then evaluate it. This assigns `0` to `x`, which is no longer a free variable
+when it is used on line 4.
+
+#### Simulating dynamic scope
+
+```r
+x <- 1
+f <- function(a) {
+  get("x", envir = parent.frame()) + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # 2
+```
+
+On line 3, we perform an explicit variable lookup for `x`, but we do so in the
+environment `parent.frame()`, which refers to the calling function's
+environment.
+
+#### Constructing an arbitrary environment
+
+```r
+x <- 1
+f <- function(a) {
+  e <- new.env()
+  e$x <- 3
+  get("x", envir = e) + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # 3
+```
+
+On line 3, we construct an empty environment, `e`. Next, on line 4, we add a
+new binding to that environment, where `x` has the value `4`. Finally, on line
+5, we perform a variable lookup in environment `e`.
+
+#### Deleting bindings
+
+```r
+x <- 1
+f <- function(a) {
+  rm("x", envir = parent.env(environment()))
+  x + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # Error in f(0) : object 'x' not found
+```
+
+On line 3, we delete the binding `x` from the given environment,
+`parent.env(environment())`. This refers to the parent (or enclosing)
+environment of the current environment of `f`, i.e. the environment where `f`
+was defined, which is the top-level environment.
+
```

Comment:
My intuition with lexical scoping is that you know the `search path` for a variable binding at the moment you compile the function/closure. Then in what part of the chain the name will be finally resolved may be late bound or also statically known. Following this intuition, I do not think any of the previous examples clearly show how to circumvent the lexical scope. I think an example that changes the `enclosing environment` of a closure at run time would make the difference

Reply:
So @JanJecmen originally had an example like:

```r
makeDynamic <- function() {
    x <- 1
    res <- function(a) a + x
    res
}

dynamic <- function(n) {
    x <- 2
    f <- makeDynamic()
    environment(f) <- environment()
    f(n)
}
dynamic(0)
```

I could replace one of the `get` examples with something like this. Is this a "counterexample" for your intuition?

Reply:
Can you get the same result by adding this line to the original `g`?

```
environment(f) <- environment()
```

Reply:
Good point, yes. I think we avoided this in `makeDynamic` because we wanted to do all the environment manipulation hacks inside `f`.